### ISSUE TITLE
Improve async/await support

### ISSUE THREAD

**Issue by paulb777** [google/gtm-session-fetcher#318]
If a fetch is in progress, via a `CheckedCancellation` wrapper on `beginFetch`, and `stopFetching` is called, the following hang occurs:
```
SWIFT TASK CONTINUATION MISUSE: beginFetch(uploadFetcher:) leaked its continuation!
2022-09-02 13:47:25.949984-0700 AppHost-FirebaseStorage-Unit-Tests[38388:17164810] SWIFT TASK CONTINUATION MISUSE: beginFetch(uploadFetcher:) leaked its continuation! 
```

From:
```
  private func beginFetch(uploadFetcher: GTMSessionUploadFetcher) async -> (Data?, Error?) {
    return await withCheckedContinuation { continuation in
      uploadFetcher.beginFetch { data, error in
        continuation.resume(returning: (data, error))
      }
    }
  }
```

I suspect that the right solution might be to add an additional stopFetching API variation that calls back its completion handlers instead of deallocating them.

More context at https://github.com/firebase/firebase-ios-sdk/pull/10161 and internally at https://groups.google.com/a/google.com/g/gtm-session-fetcher-eng/c/AFt3g1V_tsw/m/_gIYdK7bAAAJ?utm_medium=email&utm_source=footer

**Comment by mwyman** [comment]
Rather than creating an additional variant in the Objective-C layer, I wonder if this isn't best handled by creating an official Swift overlay for the `async` version of `beginFetchWithCompletionHandler:` that stores the Swift `CheckedContinuation` in a way that gets properly handled when `stopFetching` is called.

I also wonder if the preferred way to set up the method is something like `public func beginFetch() async throws -> Data` and have the both the network error and `stopFetching` case do `continuation.resume(throwing:)` with either the network error or an error type indicating the user canceled the fetch.

**Comment by paulb777** [comment]
Yes. Both of those sound good to me. The more Swifty `throws` is better than an error parameter.

However, it might not be worth managing the overhead of an overlay library unless this is just the first of other Swift-specific API improvements.
